File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    from collections import deque, defaultdict
    # 그래프를 딕셔너리로 초기화
    graph = defaultdict(list)
    for start, end in tickets:
        graph[start].append(end)
    
    # 각 출발지에서 도착지를 알파벳 순으로 정렬
    for key in graph:
        graph[key].sort(reverse=True)
    
    # BFS를 사용해 경로를 찾기
    stack = ["ICN"]
    path = []

    while stack:
        current = stack[-1]

        # 더 갈 수 있는 경로가 없으면 경로에 추가
        if not graph[current]:
            path.append(stack.pop())
        else:
            # 방문할 수 있는 경로가 있으면, 그 경로로 이동
            stack.append(graph[current].pop())
    
    # 경로를 뒤집어 반환 (DFS와 달리 거꾸로 쌓이기 때문)
    return path[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    /*
    ICN에서 출발
    [a,b]는 a->b 가능
    가능한 경로가 2개 이상일 경우 알파벳 순서가 앞서는 경로 return
    
    알고리즘 분류 : DFS 
    어떻게 풀이? 
    1. ICN : [ooo , ooo , ooo] 꼴로 저장해서 DFS
    2. 
    */
    List<String> route = new LinkedList<>();
    Map<String,PriorityQueue<String>> graph = new HashMap<>();

    public String[] solution(String[][] tickets) {
        for(String[] ticket : tickets){
            graph.computeIfAbsent(ticket[0], k -> new PriorityQueue<>()).add(ticket[1]);
        }
        
        dfs("ICN");
        return route.toArray(new String[0]);
    }
    public void dfs(String start){
        PriorityQueue<String> pq = graph.get(start);
        while(pq != null &&!pq.isEmpty() ){
            dfs(pq.poll());
        }
        route.add(0,start);
    }
}